mesop/labs/chat.py

```diff
@@ -0,0 +1,228 @@
+    ),
+  )
+
+
+@dataclass
+class Message:
+  """Chat message metadata."""
+
+  role: Literal["assistant", "user"] = "user"
+  username: str = ""
+  content: str = ""
+
+
+@me.stateclass
+class State:
+  input: str
+  output: list[Message]
+  in_progress: bool = False
+
+
+def on_input_update(State):
+  """Generic on text input handler that saves input to State using the given key.
+
+  This helper only works if you have one state instance. If use multiple state classes
+  with this helper, then only the last event handler will be stored. For more info, see
+  https://google.github.io/mesop/guides/troubleshooting/#avoid-using-closure-variables-in-event-handler.
+  """
+
+  def _on_update(e: me.InputEvent):
+    state = me.state(State)
+    setattr(state, e.key, e.value)
+
+  return _on_update
+
+
+def chat(
+  transform: Callable[[str], Generator[str, None, None] | str],
+  *,
+  title: str | None = None,
+  bot_username: str = _BOT_USERNAME_DEFAULT,
+):
+  """Creates a simple chat UI which takes in a prompt and returns a response to the
+  prompt.
+
+  This function creates event handlers for text input and output operations
+  using the provided function `transform` to process the input and generate the output.
+
+  Args:
+    transform: Function that takes in a prompt and returns a response to the prompt.
+    title: Headline text to display at the top of the UI.
+    bot_username: Name of your bot / assistant.
+  """
+  state = me.state(State)
+
+  def on_click(e: me.ClickEvent):
+    state = me.state(State)
+    if state.in_progress or not state.input:
+      return
+
+    output = state.output
+    if output is None:
+      output = []
+    output.append(Message(role=_ROLE_USER, content=state.input))
+    state.in_progress = True
+    yield
+
+    start_time = time.time()
+    output_message = transform(state.input)
+    assistant_message = Message(role=_ROLE_ASSISTANT, username=bot_username)
+    output.append(assistant_message)
+    state.output = output
+    for content in output_message:
+      assistant_message.content += content
+      if (time.time() - start_time) >= 0.5:
+        start_time = time.time()
+        yield
+    state.in_progress = False
+    yield
+
+  with me.box(style=_STYLE_APP_CONTAINER):
+    with me.box(style=_make_style_chat_ui_container(bool(title))):
+      if title:
+        me.text(title, type="headline-5", style=_STYLE_TITLE)
+      with me.box(style=_STYLE_CHAT_BOX):
+        for msg in state.output:
+          with me.box(style=_make_style_chat_bubble_wrapper(msg.role)):
+            if msg.role == _ROLE_ASSISTANT:
+              me.text(msg.username, style=_STYLE_CHAT_BUBBLE_NAME)
+            with me.box(style=_make_chat_bubble_style(msg.role)):
+              if msg.role == _ROLE_USER:
+                me.text(msg.content, style=_STYLE_CHAT_BUBBLE_PLAINTEXT)
+              else:
+                me.markdown(msg.content)
+
+      with me.box(style=_STYLE_CHAT_INPUT_BOX):
+        me.input(
```

Comment:
To clear the input, you could try doing something like:

```py
value = "" if state.clear_input else None
```

and then in `on_click`, you do state.clear_input = True, yield, state.clear_input = False, etc.

Alternatively, you could use this trick: https://google.github.io/mesop/guides/components/#resetting-a-component (it would lose the focus though).

Reply:
I tried your suggestion but it didn't work for me with the text input. I ended up using the key increment trick which worked.

I agree the downside is that it loses focus. But I think it loses focus when the user presses the send button right now. It definitely would be good to be able to control the focus somehow.

Reply:
SG this works for now, but yeah, we should add a focus API. I've filed a ticket: https://github.com/google/mesop/issues/106